### DISCUSSION TITLE
Difference between fit_gpytorch_mll_torch, fit_gpytorch_mll and fit_gpytorch_model

### DISCUSSION THREAD

**Discussion by Sam4896** [meta-pytorch/botorch#2132] [Q&A]
Hi everyone, I am very new to BoTorch and got introduced to it very recently. My application is related to Multi-task GP and BO for geometry optmization. I was going through the examples on BoTorch website. I have two questions, I would be very thankful for  your help:

1. In [MTBO](https://botorch.org/tutorials/composite_mtbo) and [HOGP](https://botorch.org/tutorials/composite_bo_with_hogp) example, fit_gpytorch_mll_torch was used to fit the GP. I went to the documentation and I was able to understand what was happening. All good till here. But then I saw [ROBO](https://botorch.org/tutorials/robust_multi_objective_bo) or [MOMF](https://botorch.org/tutorials/Multi_objective_multi_fidelity_BO) example and there fit_gpytorch_mll was used. I saw the documentation and I found that there is fit_gpytorch_model also. What is difference between the three? Also, when to use the each one of them?

2. A different question but related to my topic. I wish to use GP and BO for optimization of a certain 1-D and later 2-D geometry that is influenced by some scalar inputs. The points on the geometry would have some correlation with the other points. So would it make more sense to use multi-objective BO or use an objective function, like in [composite function BO](https://botorch.org/tutorials/composite_mtbo). That objective function would take in the values from the geometry and spit out a scalar value and try to optimize that.

**Comment by saitcakmak** [comment]
Hi @Sam4896.

1. `fit_gpytorch_model` is a deprecated alias for `fit_gpytorch_mll`, which uses scipy's L-BFGS-B optimizer under the hood to train model hyper parameters. `fit_gpytorch_mll` is what we typically use for fitting exact GP models. `fit_gpytorch_mll_torch` is an alternative method that supports optimizers from PyTorch, such as SGD, Adam etc. The reason HOGP uses `fit_gpytorch_mll_torch` is that it is a very over-parameterized model, so we don't care to optimize each hyper parameter to convergence. The person who implemented HOGP found it to train to a reasonable quality faster when using Adam as the optimizer, though I don't have any results to support that argument.

2. Just so we're on the same page, I am assuming you have many outputs that you measure over the geometry. In general, we do not recommend multi-objective optimization when there are more than 3-4 outcomes. It typically becomes quite difficult for an individual to reason about many potentially conflicting outcomes, and the optimization also gets exponentially harder as the number of outcomes increases. In your case, defining a function that characterizes the geometry in a small number of outcomes and optimizing those seems like a better approach to me. So, composite BO is definitely worth a try. If defining the utility function is tricky, [BO with preference exploration](https://botorch.org/tutorials/bope) could also be a useful tool here.

**Reply by Sam4896** [reply]
Hi @saitcakmak thanks a lot for the reply.
1. So if I understood correctly, `fit_gpytorch_mll `uses L-BFGS-B optimizer and optimizes all hyperparameters to convergence. Whereas `fit_gpytorch_mll_torch` uses torch optimizers and they optimize the hyperparameters but not till convergence? Am I correct in saying that? And I tried Kronecker MTGP with `fit_gpytorch_mll` and with `fit_gpytorch_mll_torch` both, `fit_gpytorch_mll_torch` runs much faster than `fit_gpytorch_mll`. But I am not sure whose hyperparameters are better fit to the data.
2. Yes, we are on the same page regarding the question at hand. Okay I will experiment a bit with the single objective that takes in the  outputs of the MTGP and also with preference exploration.
3. In the documentation of BoTorch, it is written that multi-output and multi-task models are different. And a model might both, or one of them. I am unable to understand the difference clearly. What is difference between using ModelListGP, SingleTaskGP and KroneckerMultiTaskGP for multi output case. The input is 5-10 dimensional and output is 9 dimensional.
4. I was of the opinion that I could directly use multi-task models for the case where I wish to model various points of a geometry (assuming that there is some correlation between them). But now I am confused whether a multi-task model is suited for my case or multi-output model?
